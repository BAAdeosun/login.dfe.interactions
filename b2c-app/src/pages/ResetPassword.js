import React from 'react';
import { animateScroll } from "react-scroll";

import components from '../components';
import { POLICIES } from '../constants/policies';
import { onChange } from '../helpers/pageUpdatesHandler';

class ResetPassword extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            email: null,
            showErrors: false,
            showB2CErrors: true,
            errors: []
        }
        this.handleSubmit = this.handleSubmit.bind(this);
        this.onChange = onChange.bind(this);
    }

    handleSubmit(e) {
        e.preventDefault();
        //hide B2C errors and only show again if we are going to submit the form to B2C
        this.setState({ showB2CErrors: false });
        //update error messages
        this.state.errors.forEach((error) => {
            error.visible.text = error.current.text;
            error.visible.showSummaryText = error.current.showSummaryText;
        });
        //do something to validate and decide if we submit or show errors
        if (this.state.email) {
            //hide our validation errors and prepare to show B2C ones (in case there are any)
            this.setState({ showErrors: false, showB2CErrors: true });
            //everything is valid, set data and submit B2C form
            this.setDataAndSubmit();
        }
        else {
            //show errors in each component
            this.setState({ showErrors: true });
            //scroll to the top of the page to show the errors
            animateScroll.scrollToTop({ duration: 500 });
        }
    }

    setDataAndSubmit() {
        //retrieve all elements we will need and set their values
        let b2cEmail = document.getElementById('email');
        let b2cSubmitButton = document.getElementById('continue');

        if (b2cEmail && b2cSubmitButton) {
            b2cEmail.value = this.state.email;
            //submit B2C form
            b2cSubmitButton.click();
        }
    }

    render() {

        const aboveFormContent =
            <div>
                <components.Paragraph>You can reset your password if you've forgotten it.</components.Paragraph>
                <components.Paragraph>If you cannot remember your email address you can also retrieve it here.</components.Paragraph>
                <h3 className="govuk-heading-m" key='heading'>Reset your password</h3>
                <components.Paragraph>To reset your password we need to send an email to the address registered to your account.</components.Paragraph>
            </div>

        const formContent =
            <components.InputField
                type='email'
                inputId='email'
                inputLabel='Email address'
                onChange={this.onChange}
                errorMessagePlaceholder='email address'
                showErrors={this.state.showErrors}
                errors={this.state.errors}
            />;

        const belowFormContent =
            <components.Paragraph>
                <components.Link policy={POLICIES.FIND_EMAIL}>Can't remember your email address?</components.Link>
            </components.Paragraph>

        const title = 'Access your account';

        const pageConfig = {
            title: title,
            header: title,
            aboveFormContent: aboveFormContent,
            formContent: formContent,
            belowFormContent: belowFormContent,
            submitButtonText: 'Send email',
            submitHandler: this.handleSubmit,
            errors: this.state.errors,
            showB2CErrors: this.state.showB2CErrors
        };


        return (
            <div id="resetPassword">
                <components.PageContainer pageConfig={pageConfig} />
            </div>
        )
    }
}

export default ResetPassword;
