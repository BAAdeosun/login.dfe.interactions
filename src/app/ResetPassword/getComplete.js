'use strict';

const action = (req, res) => {
  res.render('ResetPassword/views/complete');
};

module.exports = action;