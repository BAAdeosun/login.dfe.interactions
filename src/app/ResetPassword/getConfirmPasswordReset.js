'use strict';

const action = (req, res) => {
  res.render('ResetPassword/views/confirm', {
    csrfToken: req.csrfToken(),
    email: '',
    code: '',
    validationFailed: false,
    validationMessages: {
      email: '',
      code: ''
    }
  });
};

module.exports = action;