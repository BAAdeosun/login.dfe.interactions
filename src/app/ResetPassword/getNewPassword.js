'use strict';

const getPasswordPolicy = require('./../../infrastructure/PasswordPolicy').get;

const action = (req, res) => {
  res.render('ResetPassword/views/newpassword', {
    csrfToken: req.csrfToken(),
    title: 'Reset your password',
    newPassword: '',
    confirmPassword: '',
    validationFailed: false,
    validationMessages: {},
    passwordPolicy: getPasswordPolicy(),
  });
};

module.exports = action;
