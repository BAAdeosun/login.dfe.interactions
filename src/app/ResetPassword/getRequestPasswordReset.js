'use strict';

const action = (req, res) => {
  res.render('ResetPassword/views/request', {
    csrfToken: req.csrfToken(),
    email: '',
    validationFailed: false,
  });
};

module.exports = action;