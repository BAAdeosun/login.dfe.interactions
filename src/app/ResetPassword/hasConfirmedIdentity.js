const hasConfirmedIdentity = (req, res, next) => {
  if (req.session && req.session.uid) {
    return next();
  }
  return res.redirect('confirm');
};

module.exports = hasConfirmedIdentity;