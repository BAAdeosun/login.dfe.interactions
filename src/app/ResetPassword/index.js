'use strict';

const express = require('express');
const logger = require('./../../infrastructure/logger');

const getRequestPasswordReset = require('./getRequestPasswordReset');
const postRequestPasswordReset = require('./postRequestPasswordReset');
const getConfirmPasswordReset = require('./getConfirmPasswordReset');
const postConfirmPasswordReset = require('./postConfirmPasswordReset');
const getNewPassword = require('./getNewPassword');
const postNewPassword = require('./postNewPassword');
const getComplete = require('./getComplete');

const router = express.Router({ mergeParams: true });

const registerRoutes = (csrf) => {
  logger.info('Mounting ResetPassword routes');

  router.get('/request', csrf, getRequestPasswordReset);
  router.post('/request', csrf, postRequestPasswordReset);

  router.get('/confirm', csrf, getConfirmPasswordReset);
  router.post('/confirm', csrf, postConfirmPasswordReset);

  router.get('/newpassword', csrf, getNewPassword);
  router.post('/newpassword', csrf, postNewPassword);

  router.get('/complete', csrf, getComplete);

  return router;
};

module.exports = registerRoutes;
