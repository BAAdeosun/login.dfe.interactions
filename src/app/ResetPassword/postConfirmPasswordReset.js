'use strict';

const emailValidator = require('email-validator');
const directoriesApi = require('./../../Users');
const clients = require('./../../Clients');
const userCodes = require('./../../UserCodes');
const logger = require('./../../logger');

const validate = (email, code) => {
  const messages = {};
  let failed = false;

  if (email.length === 0) {
    messages.email = 'Please enter a valid email address';
    failed = true;
  } else if (!emailValidator.validate(email)) {
    messages.email = 'Please enter a valid email address';
    failed = true;
  }

  if (code.length === 0) {
    messages.code = 'Please enter the code that was emailed to you';
    failed = true;
  }

  return {
    failed,
    messages,
  };
};

const action = async (req, res) => {
  const validationResult = validate(req.body.email, req.body.code);

  if (validationResult.failed) {
    res.render('ResetPassword/views/confirm', {
      csrfToken: req.csrfToken(),
      email: req.body.email,
      code: req.body.code,
      validationFailed: true,
      validationMessages: validationResult.messages,
    });
    return;
  }

  let userCode;
  let user;
  try{
    const client = await clients.get(req.query.clientid);
    user = await directoriesApi.find(req.body.email, client);
    userCode = await userCodes.validateCode(user.sub, req.body.code);
  } catch(e) {
    logger.info(`Error confirming password reset for ${req.body.email}`);
    logger.info(e);
  }

  if (userCode) {
    req.session.uid = user.sub;
    req.session.clientId = req.query.clientid;
    res.redirect(`/${req.params.uuid}/resetpassword/newpassword`);
    return;
  }

  validationResult.messages.code = 'The code you entered is incorrect. Please check and try again.';
  res.render('ResetPassword/views/confirm', {
    csrfToken: req.csrfToken(),
    email: req.body.email,
    code: req.body.code,
    validationFailed: true,
    validationMessages: validationResult.messages,
  });
};

module.exports = action;
