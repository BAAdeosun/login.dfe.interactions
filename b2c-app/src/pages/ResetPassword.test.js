
import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import ResetPassword from './ResetPassword';

it('renders without crashing', () => {
    shallow(<ResetPassword />);
});

it('renders correctly', () => {
    const tree = renderer
        .create(<ResetPassword />)
        .toJSON();
    expect(tree).toMatchSnapshot();
});