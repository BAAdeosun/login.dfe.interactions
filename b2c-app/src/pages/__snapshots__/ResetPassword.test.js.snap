// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  id="resetPassword"
>
  <div>
    <div
      className="govuk-width-container"
    >
      <div
        className="govuk-breadcrumbs"
      >
        <ol
          className="govuk-breadcrumbs__list"
        >
          <li
            className="govuk-breadcrumbs__list-item"
          >
            <a
              className="govuk-breadcrumbs__link"
              href="https://nationalcareers.service.gov.uk/"
            >
              Home
            </a>
          </li>
        </ol>
      </div>
      <main
        className="govuk-main-wrapper"
      >
        <div
          className="govuk-grid-row"
        >
          <div
            className="govuk-grid-column-two-thirds"
          >
            <div
              className="pageLevelErrorContainer hide"
            >
              <div
                aria-labelledby="error-summary-title"
                className="govuk-error-summary"
                data-module="govuk-error-summary"
                role="alert"
                tabIndex="-1"
              >
                <h2
                  className="govuk-error-summary__title"
                  id="error-summary-title"
                >
                  There is a problem
                </h2>
                <div
                  className="govuk-error-summary__body"
                >
                  <ul
                    className="govuk-list govuk-error-summary__list"
                    id="errorSummaryItems"
                  />
                </div>
              </div>
            </div>
            <h1
              className="govuk-heading-xl govuk-!-margin-top-6 govuk-!-margin-bottom-9"
            >
              Access your account
            </h1>
            <div>
              <p
                className="govuk-body"
              >
                You can reset your password if you've forgotten it.
              </p>
              <p
                className="govuk-body"
              >
                If you cannot remember your email address you can also retrieve it here.
              </p>
              <h3
                className="govuk-heading-m"
              >
                Reset your password
              </h3>
              <p
                className="govuk-body"
              >
                To reset your password we need to send an email to the address registered to your account.
              </p>
            </div>
            <form
              id="customForm"
              noValidate={true}
              onSubmit={[Function]}
            >
              <div
                className="govuk-form-group "
              >
                <label
                  className="govuk-label"
                  htmlFor="emailCustom"
                >
                  Email address
                </label>
                <input
                  autoComplete="email"
                  className="govuk-input govuk-!-width-one-half"
                  id="emailCustom"
                  name="email"
                  noValidate={true}
                  onChange={[Function]}
                  spellCheck="false"
                  type="email"
                />
              </div>
              <button
                className="govuk-button"
                id="preSubmit"
                type="submit"
              >
                Send email
              </button>
            </form>
            <p
              className="govuk-body"
            >
              <a
                className="govuk-link"
                href={null}
                onClick={null}
              >
                Can't remember your email address?
              </a>
            </p>
          </div>
        </div>
      </main>
    </div>
  </div>
</div>
`;
